Classes/Middleware/CookieDomainMiddleware.php

```diff
@@ -0,0 +1,60 @@
+<?php
+
+/*
+ * This file is part of the "headless" Extension for TYPO3 CMS.
+ *
+ * For the full copyright and license information, please read the
+ * LICENSE.md file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace FriendsOfTYPO3\Headless\Middleware;
+
+use FriendsOfTYPO3\Headless\Utility\UrlUtility;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Psr\Http\Server\RequestHandlerInterface;
+use Psr\Http\Server\MiddlewareInterface;
+use TYPO3\CMS\Core\Http\NormalizedParams;
+use TYPO3\CMS\Core\Log\LogManager;
+use TYPO3\CMS\Core\Site\SiteFinder;
+use TYPO3\CMS\Core\Utility\GeneralUtility;
+
+class CookieDomainMiddleware implements MiddlewareInterface
+{
+    public function process(
+        ServerRequestInterface $request,
+        RequestHandlerInterface $handler
+    ): ResponseInterface {
+        /** @var Logger $logger */
+        $logger = GeneralUtility::makeInstance(LogManager::class)->getLogger(__CLASS__);
+
+        /** @var NormalizedParams $normalizedParams */
+        $normalizedParams = $request->getAttribute('normalizedParams');
+        $requestHost = $normalizedParams->getHttpHost();
+
+        /** @var UrlUtility $urlUtility */
+        $urlUtility = GeneralUtility::makeInstance(UrlUtility::class);
+
+        /** @var SiteFinder $siteFinder */
+        $siteFinder = GeneralUtility::makeInstance(SiteFinder::class);
+        $allSites = $siteFinder->getAllSites();
+
+        foreach ($allSites as $site) {
+            $urlUtility = $urlUtility->withSite($site);
+            $base = $urlUtility->resolveKey('base');
+            $cookieDomain = $urlUtility->resolveKey('cookieDomain');
+
+            if (str_contains($base, $requestHost)) {
```

Comment:
I think we can combine both checks into one & first check if cookieDomain is set, also if both checks are met we should break loop (in case many sites) not necessary to loop for all of them